test/suite/lib.rs

```diff
@@ -268,35 +283,228 @@ pub async fn icrc2_test_supported_standards(ledger: impl LedgerEnv) -> anyhow::R
+
+    assert_balance(&ledger_env, p1_env.principal(), fee.clone()).await?;
+    assert_balance(&ledger_env, p2_env.principal(), 0).await?;
+    assert_balance(&ledger_env, p2_subaccount.clone(), 0).await?;
+
+    // Approval for a subaccount.
+    approve(
+        &p1_env,
+        ApproveArgs::approve_amount(approve_amount.clone() * 2, p2_subaccount.clone()),
+    )
+    .await??;
+
+    assert_allowance(
+        &p1_env,
+        p1_env.principal(),
+        p2_env.principal(),
+        approve_amount.clone(),
+        None,
+    )
+    .await?;
+
+    assert_allowance(
+        &p1_env,
+        p1_env.principal(),
+        p2_subaccount.clone(),
+        approve_amount.clone() * 2,
+        None,
+    )
+    .await?;
+
+    assert_balance(&ledger_env, p1_env.principal(), 0).await?;
+    assert_balance(&ledger_env, p2_env.principal(), 0).await?;
+    assert_balance(&ledger_env, p2_subaccount, 0).await?;
+
+    // Insufficient funds to pay the fee for a second approval
+    match approve(
+        &p1_env,
+        ApproveArgs::approve_amount(Nat::from(10), p2_env.principal()),
+    )
+    .await?
+    {
+        Ok(_) => {
+            return Err(anyhow::Error::msg(
+                "expected ApproveError::InsufficientFunds, got Ok result",
+            ))
+        }
+        Err(e) => match e {
+            ApproveError::InsufficientFunds { balance } => {
+                if balance != 0 {
+                    bail!("wrong balance, expected 0, got: {}", balance);
+                }
+            }
+            _ => return Err(e).context("expected ApproveError::InsufficientFunds"),
+        },
+    }
+
+    Ok(Outcome::Passed)
+}
+
+pub async fn icrc2_test_approve_expiration(ledger_env: impl LedgerEnv) -> anyhow::Result<Outcome> {
+    let fee = transfer_fee(&ledger_env).await?;
+    let initial_balance: Nat = fee.clone() * 2;
+    let p1_env = setup_test_account(&ledger_env, initial_balance.clone()).await?;
+    let p2_env = ledger_env.fork();
+    let approve_amount = Nat::from(100);
+    let now = time_nanos(&ledger_env);
+
+    // Expiration in the past
+    match approve(
+        &p1_env,
+        ApproveArgs::approve_amount(approve_amount.clone(), p2_env.principal())
+            .expires_at(now - 100),
+    )
+    .await?
+    {
+        Ok(_) => {
+            return Err(anyhow::Error::msg(
+                "expected ApproveError::Expired, got Ok result",
+            ))
+        }
+        Err(e) => match e {
+            ApproveError::Expired { .. } => {}
+            _ => return Err(e).context("expected ApproveError::Expired"),
+        },
+    }
+
+    assert_allowance(&p1_env, p1_env.principal(), p2_env.principal(), 0, None).await?;
+
+    assert_balance(&ledger_env, p1_env.principal(), initial_balance.clone()).await?;
+    assert_balance(&ledger_env, p2_env.principal(), 0).await?;
+
+    // Correct expiration in the future
+    let expiration = time_nanos(&ledger_env) + 1000000000000000;
```

Comment:
Lets use u64::Max since we do not know what the window is

Reply:
Done!